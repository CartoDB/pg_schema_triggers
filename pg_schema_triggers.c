/*
 * ProcessUtility hook and associated functions to provide additional schema
 * change events for the CREATE EVENT TRIGGER functionality.
 *
 *
 * pg_schema_triggers/pg_schema_triggers.c
 */


#include "postgres.h"
#include "fmgr.h"
#include "access/hash.h"
#include "access/xact.h"
#include "catalog/objectaccess.h"
#include "catalog/objectaddress.h"
#include "catalog/pg_class.h"
#include "catalog/pg_type.h"
#include "parser/parse_func.h"
#include "tcop/utility.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"

#include "events.h"
#include "trigger_funcs.h"


/* PG_MODULE_MAGIC must appear exactly once in the entire module. */
PG_MODULE_MAGIC;

void _PG_init(void);
void _PG_fini(void);

static ProcessUtility_hook_type old_utility_hook = NULL;
static object_access_hook_type old_objectaccess_hook = NULL;
static int stmt_createEventTrigger_before(CreateEventTrigStmt *stmt);

static void objectaccess_hook(ObjectAccessType access,
	Oid classId,
	Oid objectId,
	int subId,
	void *arg);
static void utility_hook(Node *parsetree,
	const char *queryString,
	ProcessUtilityContext context,
	ParamListInfo params,
	DestReceiver *dest,
	char *completionTag);


/*
 * Install and uninstall the hook.
 *
 * We react poorly if we discover another module has also installed a
 * ProcessUtility hook.
 */

void
_PG_init(void)
{
	if (ProcessUtility_hook != NULL)
		elog(FATAL, "a ProcessUtility hook is already installed.");
	old_utility_hook = ProcessUtility_hook;
	ProcessUtility_hook = utility_hook;

	if (object_access_hook != NULL)
		elog(FATAL, "an object_access hook is already installed.");
	old_objectaccess_hook = object_access_hook;
	object_access_hook = objectaccess_hook;
}

void
_PG_fini(void)
{
	if (ProcessUtility_hook != utility_hook)
		elog(FATAL, "hook conflict, our ProcessUtility hook has been removed.");
	ProcessUtility_hook = old_utility_hook;

	if (object_access_hook != objectaccess_hook)
		elog(FATAL, "hook conflict, our object_access hook has been removed.");
	object_access_hook = old_objectaccess_hook;
}


/*
 * The meat of the hook.
 */

static void
utility_hook(Node *parsetree,
	const char *queryString,
	ProcessUtilityContext context,
	ParamListInfo params,
	DestReceiver *dest,
	char *completionTag)
{
	int suppress;

	/* Call the *_before() function, if any. */
	switch (nodeTag(parsetree))
	{
		case T_CreateEventTrigStmt:
			suppress = stmt_createEventTrigger_before((CreateEventTrigStmt *) parsetree);
			break;
		case T_ListenStmt:
			suppress = stmt_listen_before((ListenStmt *) parsetree);
			break;
		default:
			suppress = 0;
			break;
	}

	/* Did the before() function handle the command itself? */
	if (suppress)
		return;
	standard_ProcessUtility(parsetree, queryString, context, params, dest, completionTag);

	/* Call the *_after() function, if any. */
	switch (nodeTag(parsetree))
	{
		case T_ListenStmt:
			stmt_listen_after((ListenStmt *) parsetree);
			break;
		default:
			break;
	}
}


/*
 * We abuse the OAT_* hooks (which are intended for access control frameworks
 * such as sepgsql) for getting notified on relation create, alter, and drop
 * actions.  This is more reliable than digging around in the utility command
 * parse trees, as these hooks are called from the internal functions which
 * actually update the system catalogs, regardless of how those functions were
 * invoked.
 */
static void
objectaccess_hook(ObjectAccessType access,
	Oid classId,
	Oid objectId,
	int subId,
	void *arg)
{
	ObjectAddress object;

	object.classId = classId;
	object.objectId = objectId;
	object.objectSubId = subId;

	switch (access)
	{
		case OAT_POST_CREATE:
			{
				ObjectAccessPostCreate *args = (ObjectAccessPostCreate *)arg;

				if (!args->is_internal)
					object_post_create(&object);
			}
			break;
		case OAT_POST_ALTER:
			{
				ObjectAccessPostAlter *args = (ObjectAccessPostAlter *)arg;

				if (!args->is_internal)
					object_post_alter(&object, args->auxiliary_id);
			}
			break;
		case OAT_DROP:
			{
				ObjectAccessDrop *args = (ObjectAccessDrop *)arg;

				object_drop(&object, args->dropflags);
			}
			break;
		case OAT_NAMESPACE_SEARCH:
		case OAT_FUNCTION_EXECUTE:
			/* Ignore these events. */
			break;
	}
}


/*
 * List of events that we recognize, along with the number and types of
 * arguments that are passed to the event trigger function.
 */
struct event {
	char *eventname;
};

struct event supported_events[] = {
	{"stmt.listen.before"},
	{"stmt.listen.after"},
	{"relation.create"},

	/* end of list marker */
	{NULL}
};

/*
 * Intercept CREATE EVENT TRIGGER statements with event names that we
 * recognize, and pass them to our own CreateEventTriggerEx() function.
 * Pass everything else through to ProcessUtility_standard otherwise.
 */
static int
stmt_createEventTrigger_before(CreateEventTrigStmt *stmt)
{
    Oid         funcoid;
    int			recognized = 0;
    struct event *evt;

	/*
	 * If we recognize the event name, look up the corresponding function.
	 * Otherwise, fall through to the normal CreateEventTrigger() code.
	 */
	for (evt = supported_events; evt->eventname != NULL; evt++)
	{
		if (strcmp(stmt->eventname, evt->eventname) != 0)
			continue;

		/*
		 * LookupFuncName() will raise an error if it can't find a
		 * function with the appropriate signature.
		 */
		funcoid = LookupFuncName(stmt->funcname, 0, NULL, false);
		recognized = 1;
		break;
	}
	if (!recognized)
	{
		elog(INFO, "pg_schema_triggers:  didn't recognize event name, ignoring.");
		return 0;
	}

	/* Check the trigger function's return type. */
    if (get_func_rettype(funcoid) != EVTTRIGGEROID)
        ereport(ERROR,
                (errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
                 errmsg("function \"%s\" must return type \"%s\"",
                        get_func_name(funcoid), format_type_be(EVTTRIGGEROID))));

	/* None of our events support the WHEN clause, so ensure that it is empty. */
	if (stmt->whenclause)
		ereport(ERROR,
				(errcode(ERRCODE_SYNTAX_ERROR),
				 errmsg("event \"%s\" cannot have a WHEN clause",
						stmt->eventname)));

	/* Create the event trigger. */
    CreateEventTriggerEx(stmt->eventname, stmt->trigname, funcoid);

	/* And skip the call to CreateEventTrigger(). */
	return 1;
}
