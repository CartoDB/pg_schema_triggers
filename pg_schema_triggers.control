default_version = '0.1'
comment = 'Extends CREATE EVENT TRIGGER with additional schema events.'
module_pathname = '$libdir/pg_schema_triggers.so'
superuser = true
relocatable = false
schema = schema_triggers
