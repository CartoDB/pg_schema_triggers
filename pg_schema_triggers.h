

/* pg_schema_triggers/event_trigger.c */
Oid CreateEventTriggerEx(const char *eventname, const char *trigname, Oid trigfunc);
void FireEventTriggers(const char *eventname, const char *tag);
