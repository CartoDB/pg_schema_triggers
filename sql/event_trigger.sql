-- Test that the CREATE/ALTER/DROP EVENT TRIGGER commands work as expected.

CREATE EXTENSION schema_triggers;

-- Exercise the various cases that shouldn't work.
CREATE EVENT TRIGGER wont_work ON phase_of_the_moon
	EXECUTE PROCEDURE foo();
CREATE EVENT TRIGGER wont_work ON relation_create
	EXECUTE PROCEDURE undefined_function();
CREATE EVENT TRIGGER wrong_func_result_type ON relation_create
	EXECUTE PROCEDURE pg_client_encoding();

-- Before we create a working event trigger, we need a function to execute.
CREATE FUNCTION raise_notice()
 RETURNS event_trigger
 AS $$ BEGIN RAISE NOTICE 'do_notice:  event=(%)', TG_EVENT; END; $$
 LANGUAGE plpgsql;

-- Ensure that we're not allowed to specify a WHEN clause.
CREATE EVENT TRIGGER cannot_have_when_clause ON relation_create
	WHEN tag IN ('foo')
	EXECUTE PROCEDURE raise_notice();

-- Exercise the basic event trigger DDL.
CREATE EVENT TRIGGER one ON relation_create
	EXECUTE PROCEDURE raise_notice();
CREATE EVENT TRIGGER two ON relation_alter
	EXECUTE PROCEDURE raise_notice();
CREATE TABLE foobar();
ALTER TABLE foobar ADD COLUMN a TEXT;
DROP TABLE foobar;
ALTER EVENT TRIGGER one DISABLE;
ALTER EVENT TRIGGER two DISABLE;
CREATE TABLE foobar();
ALTER TABLE foobar ADD COLUMN a TEXT;
DROP TABLE foobar;

-- Try to drop the raise_notice() function, which should fail.
DROP FUNCTION raise_notice();
DROP EVENT TRIGGER one;
DROP EVENT TRIGGER two;
-- Now try to drop the raise_notice() function again, which should succeed.
DROP FUNCTION raise_notice();
DROP EXTENSION schema_triggers;
