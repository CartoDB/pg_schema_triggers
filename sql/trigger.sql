CREATE EXTENSION schema_triggers;

-- Create an event trigger for the trigger_create event.
CREATE FUNCTION on_trigger_create()
 RETURNS event_trigger
 LANGUAGE plpgsql
 AS $$
	DECLARE
		event_info SCHEMA_TRIGGERS.TRIGGER_CREATE_EVENTINFO;
	BEGIN
		event_info := schema_triggers.get_trigger_create_eventinfo();
		RAISE NOTICE 'on_trigger_create: (tgname=''%'', tgrelid=''%'', tgenabled=''%'', is_internal=''%'')',
			(event_info.new).tgname, (event_info.new).tgrelid::REGCLASS,
			(event_info.new).tgenabled, event_info.is_internal;
		IF (event_info.new).tgname LIKE 'test_%' THEN
			RAISE EXCEPTION 'trigger name cannot begin with "test_"';
		END IF;
		IF (event_info.new).tgrelid = 'untriggered'::REGCLASS THEN
			RAISE EXCEPTION 'cannot add triggers to relation named "untriggered".';
		END IF;
	END;
$$;
CREATE EVENT TRIGGER trigcreate ON trigger_create
	EXECUTE PROCEDURE on_trigger_create();

-- Create an event trigger for the trigger_adjust event.
-- TODO

-- Create an event trigger for the trigger_rename event.
-- TODO

-- Create an event trigger for the trigger_drop event.
-- TODO

-- Create some empty tables.
CREATE TABLE foo();
CREATE TABLE foobar();
CREATE TABLE untriggered();

-- Create a do-nothing trigger function for the following tests.
CREATE FUNCTION ignore()
 RETURNS trigger
 LANGUAGE plpgsql
 AS $$ BEGIN RETURN NULL; END; $$;

-- Create some triggers on the tables.
CREATE TRIGGER foo_trig BEFORE INSERT ON foo
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER foobar_trig BEFORE UPDATE OR DELETE ON foobar
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER test_will_not_work BEFORE INSERT ON foobar
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER also_will_not_work BEFORE INSERT ON untriggered
	EXECUTE PROCEDURE ignore();

-- Rename, enable, and disable the triggers.
ALTER TABLE foo DISABLE TRIGGER foo_trig;
ALTER TRIGGER foo_trig ON foo RENAME TO foo_trig2;
ALTER TABLE foo ENABLE TRIGGER foo_trig2;
ALTER TABLE foobar ENABLE REPLICA TRIGGER foobar_trig;
ALTER TABLE foobar ENABLE ALWAYS TRIGGER foobar_trig;
ALTER TABLE foobar DISABLE TRIGGER foobar_trig;

-- Drop the triggers.
DROP TRIGGER foo_trig2 ON foo;
DROP TRIGGER foobar_trig ON foobar;

-- Create a table with an automatically-created constraint trigger (for
-- the FOREIGN KEY) and then drop the tables, which will also drop the 
-- constraint triggers.
CREATE TABLE baz(a INTEGER PRIMARY KEY, b TEXT, c BOOLEAN);
CREATE TABLE xyzzy(d INTEGER NOT NULL REFERENCES baz);
DROP TABLE baz CASCADE;

-- Clean up.
DROP EVENT TRIGGER trigcreate;
DROP FUNCTION on_trigger_create();
DROP FUNCTION ignore();
DROP EXTENSION schema_triggers;
