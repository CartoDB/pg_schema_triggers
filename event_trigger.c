/*
 * Utility functions for creating and firing event triggers.  This code is
 * mostly copied verbatim from commands/event_trigger.c, with minor tweaks to
 * allow for new event names.
 *
 * (The exported functions dealing with event triggers are too high-level for
 * our needs.  It's a shame we have to duplicate code...)
 *
 * pg_schema_triggers/event_trigger.c
 */


#include "postgres.h"
#include "fmgr.h"
#include "miscadmin.h"
#include "access/hash.h"
#include "access/htup_details.h"
#include "catalog/dependency.h"
#include "catalog/indexing.h"
#include "catalog/objectaccess.h"
#include "catalog/objectaddress.h"
#include "catalog/pg_event_trigger.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type.h"
#include "commands/trigger.h"
#include "parser/parse_func.h"
#include "tcop/utility.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"
#include "utils/rel.h"
#include "utils/syscache.h"

#include "pg_schema_triggers.h"


Oid
CreateEventTriggerEx(const char *eventname, const char *trigname, Oid trigfunc)
{
	/* Declarations from CreateEventTrigger(). */
    HeapTuple   tuple;
    Oid         funcrettype;
    Oid			evtOwner = GetUserId();

	/* Declarations from insert_event_trigger_tuple(). */
    Relation    tgrel;
    Oid         trigoid;
    HeapTuple   tgtuple;
    Datum       values[Natts_pg_event_trigger];
    bool        nulls[Natts_pg_event_trigger];
    NameData    evtnamedata,
                evteventdata;
    ObjectAddress myself,
                referenced;

    /*
     * It would be nice to allow database owners or even regular users to do
     * this, but there are obvious privilege escalation risks which would have
     * to somehow be plugged first.
     */
    if (!superuser())
        ereport(ERROR,
                (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
                 errmsg("permission denied to create event trigger \"%s\"",
                        trigname),
                 errhint("Must be superuser to create an event trigger.")));

    /*
     * Give user a nice error message if an event trigger of the same name
     * already exists.
     */
    tuple = SearchSysCache1(EVENTTRIGGERNAME, CStringGetDatum(trigname));
    if (HeapTupleIsValid(tuple))
        ereport(ERROR,
                (errcode(ERRCODE_DUPLICATE_OBJECT),
                 errmsg("event trigger \"%s\" already exists",
                        trigname)));

    /* Validate the trigger function. */
    funcrettype = get_func_rettype(trigfunc);
    if (funcrettype != EVTTRIGGEROID)
        ereport(ERROR,
                (errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
                 errmsg("function \"%s\" must return type \"event_trigger\"",
                        get_func_name(trigfunc))));

    /* Open pg_event_trigger. */
    tgrel = heap_open(EventTriggerRelationId, RowExclusiveLock);

    /* Build the new pg_event_trigger tuple. */
    memset(nulls, false, sizeof(nulls));
    namestrcpy(&evtnamedata, trigname);
    values[Anum_pg_event_trigger_evtname - 1] = NameGetDatum(&evtnamedata);
    namestrcpy(&evteventdata, eventname);
    values[Anum_pg_event_trigger_evtevent - 1] = NameGetDatum(&evteventdata);
    values[Anum_pg_event_trigger_evtowner - 1] = ObjectIdGetDatum(evtOwner);
    values[Anum_pg_event_trigger_evtfoid - 1] = ObjectIdGetDatum(trigfunc);
    values[Anum_pg_event_trigger_evtenabled - 1] =
        CharGetDatum(TRIGGER_FIRES_ON_ORIGIN);
	nulls[Anum_pg_event_trigger_evttags - 1] = true;

    /* Insert heap tuple. */
    tgtuple = heap_form_tuple(tgrel->rd_att, values, nulls);
    trigoid = simple_heap_insert(tgrel, tgtuple);
    CatalogUpdateIndexes(tgrel, tgtuple);
    heap_freetuple(tgtuple);

    /* Depend on owner. */
    recordDependencyOnOwner(EventTriggerRelationId, trigoid, evtOwner);

    /* Depend on event trigger function. */
    myself.classId = EventTriggerRelationId;
    myself.objectId = trigoid;
    myself.objectSubId = 0;
    referenced.classId = ProcedureRelationId;
    referenced.objectId = trigfunc;
    referenced.objectSubId = 0;
    recordDependencyOn(&myself, &referenced, DEPENDENCY_NORMAL);

    /* Post creation hook for new operator family */
    InvokeObjectPostCreateHook(EventTriggerRelationId, trigoid, 0);

    /* Close pg_event_trigger and return. */
    heap_close(tgrel, RowExclusiveLock);
	return trigoid;
}


/*
 * Invoke any event triggers which are registered for the given event name.
 *
 * Unfortunately this code, similar to CreateEventTriggerEx(), has to cut-and-
 * paste a bunch of stuff from commands/event_trigger.c as there is no clean
 * API for adding new events.
 */
void
InvokeEventTriggers(const char *eventname)
{


}
