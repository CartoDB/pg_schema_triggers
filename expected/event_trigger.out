-- Test that the CREATE/ALTER/DROP EVENT TRIGGER commands work as expected.
CREATE EXTENSION schema_triggers;
-- Exercise the various cases that shouldn't work.
CREATE EVENT TRIGGER wont_work ON phase_of_the_moon
	EXECUTE PROCEDURE foo();
INFO:  pg_schema_triggers:  didn't recognize event name, ignoring.
ERROR:  unrecognized event name "phase_of_the_moon"
CREATE EVENT TRIGGER wont_work ON relation_create
	EXECUTE PROCEDURE undefined_function();
ERROR:  function undefined_function() does not exist
CREATE EVENT TRIGGER wrong_func_result_type ON relation_create
	EXECUTE PROCEDURE pg_client_encoding();
ERROR:  function "pg_client_encoding" must return type "event_trigger"
-- Before we create a working event trigger, we need a function to execute.
CREATE FUNCTION raise_notice()
 RETURNS event_trigger
 AS $$ BEGIN RAISE NOTICE 'do_notice:  event=(%)', TG_EVENT; END; $$
 LANGUAGE plpgsql;
-- Ensure that we're not allowed to specify a WHEN clause.
CREATE EVENT TRIGGER cannot_have_when_clause ON relation_create
	WHEN tag IN ('foo')
	EXECUTE PROCEDURE raise_notice();
ERROR:  event "relation_create" cannot have a WHEN clause
-- Exercise the basic event trigger DDL.
CREATE EVENT TRIGGER one ON relation_create
	EXECUTE PROCEDURE raise_notice();
CREATE EVENT TRIGGER two ON relation_alter
	EXECUTE PROCEDURE raise_notice();
CREATE TABLE foobar();
NOTICE:  do_notice:  event=(relation_create)
ALTER TABLE foobar ADD COLUMN a TEXT;
DROP TABLE foobar;
ALTER EVENT TRIGGER one DISABLE;
ALTER EVENT TRIGGER two DISABLE;
CREATE TABLE foobar();
ALTER TABLE foobar ADD COLUMN a TEXT;
DROP TABLE foobar;
-- Try to drop the raise_notice() function, which should fail.
DROP FUNCTION raise_notice();
ERROR:  cannot drop function raise_notice() because other objects depend on it
DETAIL:  event trigger one depends on function raise_notice()
event trigger two depends on function raise_notice()
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
DROP EVENT TRIGGER one;
DROP EVENT TRIGGER two;
-- Now try to drop the raise_notice() function again, which should succeed.
DROP FUNCTION raise_notice();
DROP EXTENSION schema_triggers;
